### DISCUSSION TITLE
Virtual display and Xhost

### DISCUSSION THREAD

**Discussion by adizafri2000** [adizafri2000/billscraper#6] [General]
The automation scripts will fail in the Github actions and you will encounter something like "KeyError: Display not available" or so. Even after explicitly setting the Display environment variable, you will encounter an error saying that the display cannot be opened, and maybe also a simpler one: xhost not found. To document my workaround in which I will **find out later on what steps worked and what steps are unnecessary**:

1. Add a step for setting up a headless display in the Github actions pipeline. [Link to actions repo](https://github.com/pyvista/setup-headless-display-action)
2. Add an environment variable to the pipeline: `Display=:0`
3. Install xhost in the Github actions pipeline for the Ubuntu server via `sudo apt-get install x11-xserver-utils`
4. Create a .Xauthority file in home directory using `touch ~/.Xauthority`
5. Edit /etc/pam.d/su and /etc/pam.d/sudo by appending "session optional pam_xauth.so" to the end of the file. Allow permissions using chmod if necessary ([reference](https://devicetests.com/fixing-xhost-error-ubuntu))

To review the implementation of the steps that I did, you may find the code in the python-app.yml file (Github actions) and the setup.sh file. As of doing these steps, the automation continues to run in the actions pipeline to the actual web scraping part but fails there, but this is a sign of success as I no longer encounter any xhost or display errors.

**Comment by Charlotte-br560** [comment]
Encountering display-related errors like "KeyError: Display not available" in GitHub Actions can be frustrating, but don't worry, I've got a workaround that might help!

To tackle this issue, follow these steps in your GitHub Actions pipeline:

Add a step to set up a headless display by including the necessary configurations in your pipeline. You can find an example in the actions repo.
Set an environment variable in your pipeline: Display=:0.
Install xhost in the Ubuntu server using sudo apt-get install x11-xserver-utils.
Create a .Xauthority file in the home directory by running touch ~/.Xauthority.
Edit /etc/pam.d/su and /etc/pam.d/sudo by adding session optional pam_xauth.so at the end of each file. Ensure permissions are set correctly using chmod if needed.
You can review the implementation details in the python-app.yml file for GitHub Actions and the setup.sh file.

With these steps, your automation should run smoothly without encountering xhost or display errors. Keep tweaking and refining, and you'll be scraping the web in no time!
